### PR TITLE
[fontconfig][freetds][qtwebengine] Take gperf from vcpkg

### LINKED ISSUE #24518: [fontconfig] gperf dependency from vcpkg
fontconfig require gperf.
It can be install from vcpkg.
My request is that gperf will be dependency of fontconfig. and able to detect it (add it to that PATH) and use it. 
Thank in advance.

### PR DESCRIPTION
**Pull request by talregev**
Take gperf from vcpkg tool if possible
  Fixes #24518

### CODE REVIEW

File: ports/fontconfig/portfile.cmake [hunk 1]
```diff
@@ -12,6 +12,7 @@ vcpkg_from_gitlab(
         libgetopt.patch
 )
 
+vcpkg_add_to_path("${CURRENT_HOST_INSTALLED_DIR}/tools/")
 vcpkg_find_acquire_program(GPERF)
 get_filename_component(GPERF_PATH ${GPERF} DIRECTORY)
 vcpkg_add_to_path(${GPERF_PATH})
```
Comment (Neumann-A) [inline ports/fontconfig/portfile.cmake:18]: This is not how it is done. If you want to use the port `gperf` declare it as a `host` dependency. Maybe introduce `vcpkg-tool-gperf` but you probably will be using an overlay.

Comment (talregev) [reply ports/fontconfig/portfile.cmake:18]: There is a discussion on the issue that most of the people/users doesn't want to install gperf from vcpkg, and want to install via manager app. But if the gperf is not found on the system and do found on vcpkg. What is the correct way to do it? 
Show me the way.

Comment (Neumann-A) [reply ports/fontconfig/portfile.cmake:18]: > What is the correct way to do it?

As I said: "you probably will be using an overlay".

The change here would introduce a build order dependency where the behavior changes dependent on what is installed. That is not ok.

Comment (talregev) [reply ports/fontconfig/portfile.cmake:18]: I want to change the port here. Can I change the gperf order? Also it shouldn't be vcpkg first tool?

Comment (dg0yt) [reply ports/fontconfig/portfile.cmake:18]: > The change here would introduce a build order dependency where the behavior changes dependent on what is installed. That is not ok.

However, the full story is that `scripts/buildsystems/vcpkg.cmake` adds all installed tools to `CMAKE_PROGRAM_PATH`. This is not so far away from setting `PATH`. (And I did suggest to base this dependencies instead.)

**Comment by talregev** [comment]
@BillyONeal What do you think?

File: ports/fontconfig/portfile.cmake (hunk 1 above)
Comment (ras0219-msft) [reply ports/fontconfig/portfile.cmake:18]: In my codespaces Linux VM, gperf takes 6s to install. On my Windows laptop, it takes 18 seconds. I think adding this as a vcpkg-internal dependency by default is the right answer; users that want to use it from the system can use an empty overlay.

Comment (talregev) [reply ports/fontconfig/portfile.cmake:18]: @ras0219-msft should I add gpref as a dependency to this PR?

**Comment by talregev** [comment]
@BillyONeal @ras0219-msft Any news from team discussion?

File: ports/fontconfig/portfile.cmake (hunk 1 above)
Comment (vicroms) [reply ports/fontconfig/portfile.cmake:18]: If I understand Robert correctly, yes; please add the `gperf` dependency

Comment (talregev) [reply ports/fontconfig/portfile.cmake:18]: > If I understand Robert correctly, yes; please add the `gperf` dependency

I added.

File: ports/fontconfig/vcpkg.json [hunk 2]
```diff
@@ -11,6 +11,10 @@
     "freetype",
     "getopt",
     "gettext",
+    {
+      "name": "gperf",
+      "platform": "!windows & !osx & !mingw"
```
Comment (BillyONeal) [inline ports/fontconfig/vcpkg.json:16]:
```suggestion
      "platform": "!windows & !osx & !mingw",
      "host": true
```

File: ports/fontconfig/portfile.cmake (hunk 1 above)
Comment (BillyONeal) [inline ports/fontconfig/portfile.cmake:18]:
```suggestion
```

File: ports/fontconfig/vcpkg.json (hunk 2 above)
Comment (BillyONeal) [reply ports/fontconfig/vcpkg.json:16]: And maybe the platform line should go?

Comment (LilyWangLL) [reply ports/fontconfig/vcpkg.json:16]: Or `"platform": "!windows & !osx & !mingw"` can instead of by "platform": "linux"?

Comment (talregev) [reply ports/fontconfig/vcpkg.json:16]: I can do Linux. I just copy from other dependency. 
Why libicov have that line?
https://github.com/microsoft/vcpkg/pull/24677/files#diff-85ba5805435adc92d50a623d7b117389155e2255b02e89ea703c0613ff66a57bL17

**Comment by ras0219-msft** [comment]
I've fixed gperf to install the executable into a subdirectory of tools -- could you please pull my changes into this branch?

```sh
git fetch https://github.com/ras0219-msft/vcpkg dev/roschuma/pr24677
git merge FETCH_HEAD
```

File: ports/fontconfig/vcpkg.json (hunk 2 above)
Comment (talregev) [reply ports/fontconfig/vcpkg.json]: @BillyONeal It done.

File: ports/fontconfig/portfile.cmake (hunk 1 above)
Comment (talregev) [reply ports/fontconfig/portfile.cmake]: done.

**Comment by talregev** [comment]
> I've fixed gperf to install the executable into a subdirectory of tools -- could you please pull my changes into this branch?
> 
> ```shell
> git fetch https://github.com/ras0219-msft/vcpkg dev/roschuma/pr24677
> git merge FETCH_HEAD
> ```

@ras0219-msft done.

**Comment by talregev** [comment]
@ras0219-msft I think ci is break because of your fix.

File: ports/freetds/portfile.cmake [hunk 3]
```diff
@@ -18,15 +18,7 @@ vcpkg_find_acquire_program(PERL)
...
-if (VCPKG_TARGET_IS_WINDOWS)
-    vcpkg_find_acquire_program(GPERF)
-    get_filename_component(GPERF_PATH ${GPERF} DIRECTORY)
-    vcpkg_add_to_path(${GPERF_PATH})
-else()
-    if (NOT EXISTS /usr/bin/gperf)
-        message(FATAL_ERROR "freetds requires gperf, these can be installed on Ubuntu systems via apt-get install gperf.")
-    endif()
-endif()
+vcpkg_add_to_path("${CURRENT_HOST_INSTALLED_DIR}/tools/gperf")
```
Comment (dg0yt) [inline ports/freetds/portfile.cmake:21]: `PREPEND` as in `fontconfig`?

File: ports/kf5codecs/portfile.cmake [hunk 4]
```diff
@@ -9,13 +9,11 @@ vcpkg_from_github(
...
 file(WRITE "${SOURCE_PATH}/.clang-format" "DisableFormat: true\nSortIncludes: false\n")
 
-find_program(GPERF_EXE NAMES gperf PATHS "${CURRENT_HOST_INSTALLED_DIR}/tools" NO_DEFAULT_PATH)
-
 vcpkg_cmake_configure(
     SOURCE_PATH "${SOURCE_PATH}"
     OPTIONS
         -DBUILD_TESTING=OFF
-        "-DGperf_EXECUTABLE=${GPERF_EXE}"
+        "-DGperf_EXECUTABLE=${CURRENT_HOST_INSTALLED_DIR}/tools/gperf${VCPKG_HOST_EXECUTABLE_SUFFIX}"
```
Comment (dg0yt) [inline ports/kf5codecs/portfile.cmake:16]: `${CURRENT_HOST_INSTALLED_DIR}/tools/gperf/gperf${VCPKG_HOST_EXECUTABLE_SUFFIX}`

(an extra `gperf/` dir).

**Review by dg0yt — CHANGES_REQUESTED** [review]

**Comment by talregev** [comment]
@dg0yt @JackBoosY @ras0219-msft 
Can you help me fix kf5codecs?

```
Installing 44/44 kf5codecs:x64-linux...
Building kf5codecs[core]:x64-linux...
-- Downloading https://github.com/KDE/kcodecs/archive/v5.89.0.tar.gz -> KDE-kcodecs-v5.89.0.tar.gz...
-- Extracting source /home/ubuntu/vcpkg/downloads/KDE-kcodecs-v5.89.0.tar.gz
-- Using source at /home/ubuntu/vcpkg/buildtrees/kf5codecs/src/v5.89.0-a370638e5e.clean
-- Configuring x64-linux-rel
-- Building x64-linux-rel
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:158 (message):
    Command failed: /snap/cmake/1088/bin/cmake --build . --config Release --target install -- -v -j2
    Working Directory: /home/ubuntu/vcpkg/buildtrees/kf5codecs/x64-linux-rel
    See logs for more information:
      /home/ubuntu/vcpkg/buildtrees/kf5codecs/install-x64-linux-rel-out.log

Call Stack (most recent call first):
  installed/x64-linux/share/vcpkg-cmake/vcpkg_cmake_build.cmake:112 (vcpkg_execute_build_process)
  installed/x64-linux/share/vcpkg-cmake/vcpkg_cmake_install.cmake:41 (vcpkg_cmake_build)
  ports/kf5codecs/portfile.cmake:19 (vcpkg_cmake_install)
  scripts/ports.cmake:146 (include)
```
error:
`/bin/sh: 1: /home/ubuntu/vcpkg/installed/x64-linux/tools/gperf: Permission denied`
[install-x64-linux-rel-out.log](https://github.com/microsoft/vcpkg/files/8789589/install-x64-linux-rel-out.log)

File: ports/kf5codecs/portfile.cmake (hunk 4 above)
Comment (talregev) [reply ports/kf5codecs/portfile.cmake:16]: I think this is the bug. I will try to change that.

File: ports/freetds/portfile.cmake (hunk 3 above)
Comment (talregev) [reply ports/freetds/portfile.cmake:21]: @dg0yt Can you elaborate?

File: ports/kf5codecs/portfile.cmake (hunk 4 above)
Comment (talregev) [reply ports/kf5codecs/portfile.cmake]: I fix the bug. thank you.

File: ports/freetds/portfile.cmake (hunk 3 above)
Comment (talregev) [reply ports/freetds/portfile.cmake]: done.

**Comment by talregev** [comment]
@BillyONeal @ras0219-msft @JackBoosY 
Ready for review.

**Review by BillyONeal — APPROVED** [review]

**Review by JackBoosY — APPROVED** [review]

**Review by ras0219-msft — APPROVED** [review]
Thanks!

**Review by ras0219-msft — CHANGES_REQUESTED** [review]
Sorry, one final thing. LGTM after that fix!

File: ports/qtwebengine/vcpkg.json [hunk 5]
```diff
@@ -2,7 +2,7 @@
   "$comment": "x86-windows is not within the upstream support matrix of Qt6",
   "name": "qtwebengine",
   "version": "6.3.0",
-  "port-version": 1,
+  "port-version": 2,
```
[the remark this answers is not in the document]
Comment (talregev) [reply ports/qtwebengine/vcpkg.json:5]: ok.

[the remark this answers is not in the document]
Comment (talregev) [reply ports/qtwebengine/vcpkg.json:5]: done.

**Comment by talregev** [comment]
@ras0219-msft Ready for review.

**Review by ras0219-msft — APPROVED** [review]

**Pull request merged**